File: backend/app/schemas/inventario.py
from pydantic import BaseModel, Field
from typing import Optional, List
from datetime import datetime
from decimal import Decimal


# ─────────────────────────────────────────────────────
# CUENTA CONTABLE INVENTARIO
# ─────────────────────────────────────────────────────
class CuentaContableInventarioBase(BaseModel):
    codigo_cuenta: str = Field(..., max_length=35)
    nombre_cuenta: str = Field(..., max_length=60)
    tipo_movimiento: int = Field(..., description="1=solo movimiento, 2=nivel padre")
    nivel: int = Field(..., description="3=titulo, 6=grupo, 9=cuenta movible")
    cuenta_gasto: Optional[str] = Field(None, max_length=35)

class CuentaContableInventarioCrear(CuentaContableInventarioBase):
    pass

class CuentaContableInventario(CuentaContableInventarioBase):
    id: int
    class Config:
        from_attributes = True


# ─────────────────────────────────────────────────────
# DESTINO
# ─────────────────────────────────────────────────────
class DestinoBase(BaseModel):
    nombre: str = Field(..., max_length=100)
    activo: bool = True

class DestinoCrear(DestinoBase):
    pass

class DestinoActualizar(BaseModel):
    nombre: Optional[str] = Field(None, max_length=100)
    activo: Optional[bool] = None

class Destino(DestinoBase):
    id: int
    id_origen_mysql: Optional[int] = None
    class Config:
        from_attributes = True


# ─────────────────────────────────────────────────────
# ENCARGADO
# ─────────────────────────────────────────────────────
class EncargadoBase(BaseModel):
    nombre: str = Field(..., max_length=100)
    cedula: Optional[str] = Field(None, max_length=10)
    activo: bool = True

class EncargadoCrear(EncargadoBase):
    pass

class EncargadoActualizar(BaseModel):
    nombre: Optional[str] = Field(None, max_length=100)
    cedula: Optional[str] = Field(None, max_length=10)
    activo: Optional[bool] = None

class Encargado(EncargadoBase):
    id: int
    id_origen_mysql: Optional[int] = None
    class Config:
        from_attributes = True


# ─────────────────────────────────────────────────────
# PROVEEDOR
# ─────────────────────────────────────────────────────
class ProveedorBase(BaseModel):
    nombre: str = Field(..., max_length=100)
    ruc_cedula: Optional[str] = Field(None, max_length=13)
    direccion: Optional[str] = Field(None, max_length=100)
    telefono: Optional[str] = Field(None, max_length=15)
    contacto: Optional[str] = Field(None, max_length=100)
    activo: bool = True

class ProveedorCrear(ProveedorBase):
    pass

class ProveedorActualizar(BaseModel):
    nombre: Optional[str] = Field(None, max_length=100)
    ruc_cedula: Optional[str] = Field(None, max_length=13)
    direccion: Optional[str] = Field(None, max_length=100)
    telefono: Optional[str] = Field(None, max_length=15)
    contacto: Optional[str] = Field(None, max_length=100)
    activo: Optional[bool] = None

class Proveedor(ProveedorBase):
    id: int
    id_origen_mysql: Optional[int] = None
    class Config:
        from_attributes = True


# ─────────────────────────────────────────────────────
# UNIDAD DE GESTIÓN
# ─────────────────────────────────────────────────────
class UnidadGestionBase(BaseModel):
    codigo: str = Field(..., max_length=10)
    nombre: str = Field(..., max_length=100)
    puede_mover: bool = True

class UnidadGestionCrear(UnidadGestionBase):
    pass

class UnidadGestion(UnidadGestionBase):
    id: int
    id_origen_mysql: Optional[int] = None
    class Config:
        from_attributes = True


# ─────────────────────────────────────────────────────
# ARTÍCULO INVENTARIO
# ─────────────────────────────────────────────────────
class ArticuloInventarioBase(BaseModel):
    codigo_articulo: str = Field(..., max_length=8)
    nombre: str = Field(..., max_length=100)
    descripcion_extendida: Optional[str] = Field(None, max_length=254)
    codigo_cuenta: str = Field(..., max_length=35)
    unidad_medida: Optional[str] = Field(None, max_length=10)
    existencia_inicial: Optional[Decimal] = None
    costo_inicial: Optional[Decimal] = None
    existencia_actual: Optional[Decimal] = None
    costo_actual: Decimal = Decimal("0")
    valor_total: Optional[Decimal] = None
    stock_minimo: Optional[Decimal] = None
    stock_maximo: Optional[Decimal] = None
    codigo_barras: Optional[str] = Field(None, max_length=45)
    anno_fiscal: int = Field(..., ge=2021, le=2025)
    activo: bool = True

class ArticuloInventarioCrear(ArticuloInventarioBase):
    pass

class ArticuloInventarioActualizar(BaseModel):
    nombre: Optional[str] = Field(None, max_length=100)
    descripcion_extendida: Optional[str] = Field(None, max_length=254)
    unidad_medida: Optional[str] = Field(None, max_length=10)
    existencia_actual: Optional[Decimal] = None
    costo_actual: Optional[Decimal] = None
    stock_minimo: Optional[Decimal] = None
    stock_maximo: Optional[Decimal] = None
    activo: Optional[bool] = None

class ArticuloInventario(ArticuloInventarioBase):
    id: int
    id_origen_mysql: Optional[int] = None
    creado_en: Optional[datetime] = None
    modificado_en: Optional[datetime] = None
    class Config:
        from_attributes = True


# ─────────────────────────────────────────────────────
# MOVIMIENTO
# ─────────────────────────────────────────────────────
class MovimientoBase(BaseModel):
    numero_movimiento: int
    fecha: datetime
    tipo_movimiento: str = Field(..., description="ENTRADA o SALIDA")
    subtipo: Optional[str] = Field(None, max_length=1, description="R=Reversal")
    costo_total: Optional[Decimal] = None
    numero_factura: Optional[str] = Field(None, max_length=20)
    fecha_factura: Optional[datetime] = None
    comprobante_egreso: Optional[str] = Field(None, max_length=20)
    observacion: Optional[str] = Field(None, max_length=254)
    numero_entrada_ref: Optional[int] = None
    aprobado: Optional[int] = 0
    anexo: Optional[str] = Field(None, max_length=20)
    anno_fiscal: int = Field(..., ge=2021, le=2025)
    id_proveedor: Optional[int] = None
    id_destino: Optional[int] = None
    id_encargado: Optional[int] = None

class MovimientoCrear(MovimientoBase):
    pass

class Movimiento(MovimientoBase):
    id: int
    id_origen_mysql: Optional[int] = None
    creado_en: Optional[datetime] = None
    proveedor: Optional[Proveedor] = None
    destino: Optional[Destino] = None
    encargado: Optional[Encargado] = None
    class Config:
        from_attributes = True


# ─────────────────────────────────────────────────────
# MOVIMIENTO DETALLE (KARDEX)
# ─────────────────────────────────────────────────────
class MovimientoDetalleBase(BaseModel):
    id_movimiento: int
    id_articulo: int
    fecha: datetime
    tipo_movimiento: str = Field(..., description="ENTRADA o SALIDA")
    cantidad: Decimal
    costo_unitario: Decimal
    total_linea: Optional[Decimal] = None
    costo_promedio: Optional[Decimal] = None
    anno_fiscal: int = Field(..., ge=2021, le=2025)

class MovimientoDetalleCrear(MovimientoDetalleBase):
    pass

class MovimientoDetalle(MovimientoDetalleBase):
    id: int
    id_origen_mysql: Optional[int] = None
    articulo: Optional[ArticuloInventario] = None
    class Config:
        from_attributes = True


# ─────────────────────────────────────────────────────
# SCHEMAS DE CONSULTA / LISTADO
# ─────────────────────────────────────────────────────
class ArticuloListado(BaseModel):
    id: int
    codigo_articulo: str
    nombre: str
    unidad_medida: Optional[str] = None
    existencia_actual: Optional[Decimal] = None
    costo_actual: Decimal
    valor_total: Optional[Decimal] = None
    codigo_cuenta: str
    anno_fiscal: int
    activo: bool
    class Config:
        from_attributes = True


class MovimientoListado(BaseModel):
    id: int
    numero_movimiento: int
    fecha: datetime
    tipo_movimiento: str
    costo_total: Optional[Decimal] = None
    observacion: Optional[str] = None
    anno_fiscal: int
    proveedor: Optional[str] = None  # nombre
    destino: Optional[str] = None    # nombre
    encargado: Optional[str] = None  # nombre
    class Config:
        from_attributes = True


class ResumenAnualInventario(BaseModel):
    anno_fiscal: int
    total_articulos: int
    total_movimientos: int
    total_lineas_kardex: int
    valor_inventario_total: Optional[Decimal] = None

File: frontend/src/types/inventario.ts
// ============================================================
// Tipos TypeScript — Módulo Inventario / Existencias
// Nomenclatura en español conforme a Regla Arquitectónica 2
// ============================================================

// ── Tipos de movimiento ──────────────────────────────────────
export type ENTRADA = 'ENTRADA';
export type SALIDA = 'SALIDA';

// ── Cuenta Contable de Inventario ───────────────────────────
export interface CuentaContableInventario {
  id: number;
  codigo_cuenta: string;
  nombre_cuenta: string;
  tipo_movimiento: number;
  nivel: number;
  cuenta_gasto?: string;
}

// ── Destino ──────────────────────────────────────────────────
export interface Destino {
  id: number;
  nombre: string;
  activo: boolean;
  id_origen_mysql?: number;
}

// ── Encargado ────────────────────────────────────────────────
export interface Encargado {
  id: number;
  nombre: string;
  cedula?: string;
  activo: boolean;
  id_origen_mysql?: number;
}

// ── Proveedor ────────────────────────────────────────────────
export interface Proveedor {
  id: number;
  nombre: string;
  ruc_cedula?: string;
  direccion?: string;
  telefono?: string;
  contacto?: string;
  activo: boolean;
  id_origen_mysql?: number;
}

// ── Unidad de Gestión ────────────────────────────────────────
export interface UnidadGestion {
  id: number;
  codigo: string;
  nombre: string;
  puede_mover: boolean;
  id_origen_mysql?: number;
}

// ── Artículo (listado resumido) ──────────────────────────────
export interface ArticuloListado {
  id: number;
  codigo_articulo: string;
  nombre: string;
  unidad_medida?: string;
  existencia_actual?: number;
  costo_actual: number;
  valor_total?: number;
  codigo_cuenta: string;
  anno_fiscal: number;
  activo: boolean;
}

// ── Artículo (detalle completo) ──────────────────────────────
export interface Articulo {
  id: number;
  codigo_articulo: string;
  nombre: string;
  descripcion_extendida?: string;
  codigo_cuenta: string;
  unidad_medida?: string;
  existencia_inicial?: number;
  costo_inicial?: number;
  existencia_actual?: number;
  costo_actual: number;
  valor_total?: number;
  stock_minimo?: number;
  stock_maximo?: number;
  codigo_barras?: string;
  anno_fiscal: number;
  activo: boolean;
  usuario_crea?: string;
  creado_en?: string;
  modificado_en?: string;
  id_origen_mysql?: number;
}

// ── Movimiento (listado resumido) ────────────────────────────
export interface MovimientoListado {
  id: number;
  numero_movimiento: number;
  fecha: string;
  tipo_movimiento: ENTRADA | SALIDA;
  costo_total?: number;
  observacion?: string;
  anno_fiscal: number;
  proveedor?: string;
  destino?: string;
  encargado?: string;
}

// ── Movimiento (detalle completo) ────────────────────────────
export interface Movimiento {
  id: number;
  numero_movimiento: number;
  fecha: string;
  tipo_movimiento: ENTRADA | SALIDA;
  subtipo?: string;
  costo_total?: number;
  numero_factura?: string;
  fecha_factura?: string;
  comprobante_egreso?: string;
  observacion?: string;
  numero_entrada_ref?: number;
  aprobado?: number;
  anexo?: string;
  anno_fiscal: number;
  id_proveedor?: number;
  id_destino?: number;
  id_encargado?: number;
  id_origen_mysql?: number;
  creado_en?: string;
  proveedor?: Proveedor;
  destino?: Destino;
  encargado?: Encargado;
}

// ── Línea de Kardex ──────────────────────────────────────────
export interface LineaKardex {
  id: number;
  id_movimiento: number;
  id_articulo: number;
  fecha: string;
  tipo_movimiento: ENTRADA | SALIDA;
  cantidad: number;
  costo_unitario: number;
  total_linea?: number;
  costo_promedio?: number;
  anno_fiscal: number;
  id_origen_mysql?: number;
  articulo?: Articulo;
}

// ── Resumen Anual ────────────────────────────────────────────
export interface ResumenAnualInventario {
  anno_fiscal: number;
  total_articulos: number;
  total_movimientos: number;
  total_lineas_kardex: number;
  valor_inventario_total?: number;
}
